docs: Enhance Bootstrap node deployment docs (#2178)
scality/metalk8s: docs/installation/bootstrap.rst

      root@bootstrap $ mkdir /etc/metalk8s

#. Create the :file:`/etc/metalk8s/bootstrap.yaml` file.
   This file contains initial configuration settings which are mandatory for

>    settings up a MetalK8s :term:`Bootstrap node`.

Reviewer: Whoops, missed that one:

```suggestion
   setting up a MetalK8s :term:`Bootstrap node`.
```

      archives:
        - <path-to-metalk8s-iso>

The ``networks`` field specifies a range of IP addresses for pod networks and
services in the Controlplane and Workloadplane written in CIDR notation.
Note that after the Bootstrap node is installed, these values cannot be changed
automatically.
In production clusters, we advise users to choose a
subnet range of atleast `/16` in order to have sufficient IP addresses for Pods
and services keeping in mind that users infrastructure will grow in the near
future. For example;

   .. code-block:: shell

      networks:
        controlPlane: 10.200.0.0/16
        workloadPlane: 10.200.0.0/16

The ``proxies`` field can be omitted if there is no proxy to configure.
The 2 entries ``http`` and ``https`` are used to configure the containerd
daemon proxy to fetch extra container images from outstide the MetalK8s
cluster.
The ``no_proxy`` entry specifies IPs that should be excluded from proxying,
it must be a list of hosts, IP addresses or IP ranges in CIDR format.
For example;

   .. code-block:: shell

>       NO_PROXY="localhost,127.0.0.1,10.10.0.0/16,192.168.0.0/16"

Reviewer 2: This seems confusing: not in YAML-format, and not using the same key name (caps vs. lower). Also, the YAML value must be a list (type), not a comma-separate string.

services in the Controlplane and Workloadplane written in CIDR notation.
Note that after the Bootstrap node is installed, these values cannot be changed
automatically.
In production clusters, we advise users to choose a
subnet range of atleast `/16` in order to have sufficient IP addresses for Pods
and services keeping in mind that users infrastructure will grow in the near
future. For example;

   .. code-block:: shell

      networks:
        controlPlane: 10.200.0.0/16
        workloadPlane: 10.200.0.0/16

The ``proxies`` field can be omitted if there is no proxy to configure.
The 2 entries ``http`` and ``https`` are used to configure the containerd
daemon proxy to fetch extra container images from outstide the MetalK8s
cluster.
The ``no_proxy`` entry specifies IPs that should be excluded from proxying,
it must be a list of hosts, IP addresses or IP ranges in CIDR format.
For example;

   .. code-block:: shell

      no_proxy:
        - localhost
        - 127.0.0.1
        - 10.10.0.0/16
        - 192.168.0.0/16

The ``archives`` field is a list of absolute paths to MetalK8s ISO files. When
the bootstrap script is executed, these ISOs are automatically mounted and the

> system is configured to re-mount them automatically after a reboot.

Reviewer 3: Is the bootstrap script really mounting those iso when we run it? isn't it a manual step to perform? and when we execute the bootstrap script, we update system configuration to mount the iso automatically after a reboot?

Author: Rephrased and fixed!

      archives:
        - <path-to-metalk8s-iso>

The ``networks`` field specifies a range of IP addresses written in CIDR
notation for the various fields.

In production clusters, we advise users to choose a

> subnet range of atleast `/16` in order to have sufficient IP addresses keeping

Reviewer: `/16` is definitely **not** a requirement for control/workload plane networks: those are real IPs, so as long as you have all your nodes in them, it's enough (and 5 nodes fit in a `/28` if needed).
`/16` is however recommended for internal, virtual networks that are used for Pod and Service ClusterIPs.

> in mind that users infrastructure will grow in the near future.

Reviewer: `users infrastructure` doesn't mean much to me. The recommendation makes sense however, so maybe rephrasing it to something like:

> For production clusters, we advise users to anticipate future expansions and use sufficiently large networks for control plane and workload plane. A `/24` should be enough. Note that aggregation of subnet segments is not yet achievable with this configuration file.

      The ``controlPlane`` and ``workloadPlane`` entries are **mandatory**.

>       These values specify the range of IP addresses that will be attributed to

Reviewer: Not `will be attributed`: those networks are the ranges of IP addresses used at the host level for each member of the cluster. It could also make sense to mention that **all** nodes in the cluster **must** have an interface connected to each of those networks (CP and WP) - this may be the same interface if we only use a single network range for both.

>       controlPlane and workloadPlane nodes.

Reviewer: ```suggestion
      control plane and workload plane nodes.
```

      .. code-block:: shell

            networks:
              controlPlane: 10.200.0.0/16
              workloadPlane: 10.200.0.0/16

      The ``pods`` and ``services`` fields at not mandatory but can be set

>       based on users requirements.

Reviewer: Maybe suggest why one would want to change these (`users requirements` is too vague IMHO). Something like:

> The ``pods`` and ``services`` fields are not mandatory, though can be changed to match the constraints of existing networking infrastructure (e.g. if all or part of these default subnets is already routed).

      During installation, by default ``pods`` and ``services`` are set to the
      following values below if omitted. This is done in order that pods and
      services within the cluster obtain unique IP addresses and avoid any

>       conflicts.

Reviewer: The explanation lacks a useful disclaimer: those are internal, virtual networks. Nothing has to be done on the infrastructure side, as our cluster will handle configuration of the routing tables and allocation pools on its own (at the `iptables` level). We just need to make sure that they don't conflict with existing, routed networks.

      archives:
        - <path-to-metalk8s-iso>

The ``networks`` field specifies a range of IP addresses written in CIDR
notation for the various fields.

For production clusters, we advise users to anticipate future expansions and
use sufficiently large networks for control plane and workload plane.

      The ``controlPlane`` and ``workloadPlane`` entries are **mandatory**.
      These values specify the range of IP addresses that will be attributed to
      control plane and workload plane nodes.

>       .. code-block:: shell

Reviewer: For this one, and others below, why not use `.. code-block:: yaml`?

            networks:
              controlPlane: 10.200.1.0/28
              workloadPlane: 10.200.1.0/28

      The ``pods`` and ``services`` fields are not mandatory, though can be
      changed to match the constraints of existing networking infrastructure

>       (for example, if all or part of these default subnets is already routed).

Reviewer: :+1:

      During installation, by default ``pods`` and ``services`` are set to the
      following values below if omitted.

      All nodes within the cluster **must** connect to both the control plane

>       and workload networks. If the same network range is chosen for both the

Reviewer: ```suggestion
      and workload plane networks. If the same network range is chosen for both the
```

>       control plane and workload plane networks then the same interface maybe

Reviewer: ```suggestion
      control plane and workload plane networks then the same interface may be
```

>       used.

Reviewer: I think this paragraph should be moved just after the `controlPlane`/`workloadPlane` example, for clarity. Here, we're talking about `pods`/`services`, so it's a bit confusing.

Author: true

    destination fields of IP packets to correspond to the MAC address(es))
    :ref:`IP-in-IP needs to be enabled<enable IP-in-IP>`.

Validate the install
^^^^^^^^^^^^^^^^^^^^

> - Check if all :term:`Pods <Pod>` on the Bootstrap node are in the

Reviewer 2: If we don't mention provisioning storage for Prometheus and such before this, then they *can't* be all in `Running` state.

    destination fields of IP packets to correspond to the MAC address(es)),
    :ref:`IP-in-IP needs to be enabled<enable IP-in-IP>`.

Validate the install
^^^^^^^^^^^^^^^^^^^^
- Check that all :term:`Pods <Pod>` on the Bootstrap node are in the

>   ``Running`` state. Note that Prometheus and Alertmanager pods will remain in

Reviewer: From what I've seen in @lucieleonard 's PR, user-facing values such as an object's `state` are shown as bold, not monospace:

```suggestion
  **Running** state. Note that Prometheus and Alertmanager pods will remain in
```

>   a ``Pending`` state until their respective persistent storage volumes are

Reviewer: ```suggestion
  a **Pending** state until their respective persistent storage volumes are
```

   metalk8s-monitoring   prometheus-operator-prometheus-node-exporter-rb969        1/1     Running   0          6m17s   10.200.3.152     bootstrap   <none>           <none>
   metalk8s-monitoring   prometheus-prometheus-operator-prometheus-0               0/3     Pending   0          5m50s   <none>           <none>      <none>           <none>
   metalk8s-ui           metalk8s-ui-6f74ff4bc-fgk86                               1/1     Running   0          6m4s    10.233.220.139   bootstrap   <none>           <none>

- From the console output above, :term:`Prometheus` and :term:`Alertmanager`
  pods are in a ``Pending`` state because their respective persistent
  storage volumes need to be provisioned. To provision these persistent storage
  volumes, follow :ref:`this procedure <Provision Prometheus storage>`.

- Check that you can access the MetalK8s GUI after the
  :ref:`installation <Bootstrap installation>` is completed by following
  :ref:`this procedure <quickstart-services-admin-ui>`.

  .. note::

     The MetalK8s GUI will only become accessible when persistent storage
     volumes for both Prometheus and Alertmanager have been successfully

>      provisioned.

Reviewer: That's not true anymore, UI is accessible, but Monitoring features will be missing.

   metalk8s-monitoring   prometheus-prometheus-operator-prometheus-0               0/3     Pending   0          5m50s   <none>           <none>      <none>           <none>
   metalk8s-ui           metalk8s-ui-6f74ff4bc-fgk86                               1/1     Running   0          6m4s    10.233.220.139   bootstrap   <none>           <none>

- From the console output above, :term:`Prometheus` and :term:`Alertmanager`
  pods are in a ``Pending`` state because their respective persistent
  storage volumes need to be provisioned. To provision these persistent storage
  volumes, follow :ref:`this procedure <Provision Prometheus storage>`.

- Check that you can access the MetalK8s GUI after the
  :ref:`installation <Bootstrap installation>` is completed by following
  :ref:`this procedure <quickstart-services-admin-ui>`.

- At this stage, the MetalK8s GUI should be up and ready for you to
  explore.

  .. note::

     Monitoring through the MetalK8s GUI will not be available until persistent
     storage volumes for both Prometheus and Alertmanager have been successfully
     provisioned.

> - If you encouter an error during **installation** or have difficulties

Reviewer: Not sure the bold font is useful here.

>    metalk8s-monitoring   prometheus-k8s-1                                 3/3       Running   1          3m        10.233.132.78   bootstrap   <none>

Reviewer 3: Ideally use the right version of K8S instead of 1.11.7

   metalk8s-monitoring   prometheus-operator-64477d4bff-xxjw2             1/1       Running   0          4m        10.233.132.68   bootstrap   <none>

- Check and verify that you can access the MetalK8s GUI after the
  :ref:`installation <Bootstrap installation>` is completed by following
  :ref:`this procedure <quickstart-services-admin-ui>`.

> - At this stage, the MetalK8s GUI should be up and ready for you to

Reviewer 3: even if prom and am PV have not been provisioned?

Author: fixed
